### DISCUSSION TITLE
General Chat

### DISCUSSION THREAD

**Discussion by giluriel** [8VIM/8VIM#10] [General]
Hi Ravi,
I cannot express my happyness when i ran into this page.
I was using 8pen until it dissapeared and since than I was looking for it over and over...
I'm happy you are into it.
Questions:
Is it based on michael's code (just curiouesity)? I lost contact with him too..
More important:
Do you need help with dev or testing?
How do I compile an apk?

Thabk you
Gil Uriel
Israel

**Comment by flide** [comment]
Well Gil, I can't express how happy I am to know someone likes this project. I thought I am alone in the world who wanted to use this method of typing.
I don't know who Michael is, so... this code was developed exclusively from scratch.

I have never tried this on another machine, but you should be simply able to clone this repo, import it to android studio and build it to generate the apk.

As far as help is concerned... HELL YEAH!! I need some. take a look at the open issues, maybe you can fix one?

**Comment by geofrey** [comment]
There are still occasional comments on Twitter asking when 8pen is coming back. 'Michael' is [@michaelfester](https://github.com/michaelfester) and he's active on other projects but never responds off-topic so I have a feeling there are lawyers (or bankers, I don't know which is worse) involved and we should just raise a glass to the memory of 8pen - may it rest in peace - and move on. I even tried to PayPal the official support account five dollars to get their attention but it timed out and the money came back.

Anyway I'm also quite pleased to find this project. Got it installed on my phone and I make a point of switching IMEs from time to time to stay in practice.

**Comment by giluriel** [comment]
> Well Gil, I can't express how happy I am to know someone likes this project. I thought I am alone in the world who wanted to use this method of typing.
.......... 
> As far as help is concerned... HELL YEAH!! I need some. take a look at the open issues, maybe you can fix one?

Ohh good ... 
Since I ran into 8Pen - I couldn't stop using it for a long time afterward I used old AKPs i gut extracted from my old phone but somewhere along the way - I lost the Hebrew addition and I had to give it up.
but - I never stopped looking for it.
So happy you took it to your hands :)

I'll dive into your code and try to do my best :)
once again - many thanks

**Comment by AlecLangford** [comment]
Gill if you do what I did, (you can use my fork if you want)
configure aws code build as per here
https://aws.amazon.com/blogs/mobile/automatically-build-your-android-app-with-aws-codebuild/

Just look at the build spec in my fork
https://github.com/AlecLangford/VI8

this won't exceed the aws free tier unless you're doing it on some automated basis

**Comment by akrivka** [comment]
I would like to add my appreciation! I was so frustrated using an old typewriter layout on a phone. I haven't used 8pen but Vi8 rocks, I'm already getting comfortably fast. :) 

I started porting Vi8 to my language (Czech) and came across a problem, that we have 42 letters in total. So it's eithter adding accents like you do on a normal keyboard (first ' then a to make á) or secondary layout that could be toggled by dragging quickly across either (preferably lower right) line. 

I've also had some other ideas... Are you still working on the project? I think it has great potential.

**Comment by akrivka** [comment]
Okay, so I've had a look at your way of entering Esperanto characters and I think it's much better! I will try to implement this with Czech language. :)

**Comment by flide** [comment]
Hey aidam38, nice to see new non-8pen users to join the Vi8 community (almost 100 strong now :1st_place_medal: ).
Yes, the way I enter Esperanto characters is one way, but it works for esperanto because there are only a handful of characters and frequency of them is also not so high, however, if you are willing to do the effort what you might be able to do is create a alternative board with 5 sectors rather than 4. That way you would be able to have 50 characters to enter on the board rather than 32 characters. I suggest this because that was my original idea when I was thinking of implementing Hindi to Vi8.

And to confirm, yes I work on this project, not as regularly as I would like though.

**Comment by mentalisttraceur** [comment]
Typing this using vi8 right now. Can't even begin to express how glad I am that someone started an open source clone of this.

Keep up the great work!

I might try to follow up with some separate issues for discussion of features and/or pull requests later.

**Comment by WuerfelDev** [comment]
Hey there, since i heard about 8pen i was excited. I searched and found this project some time ago. I cloned it and build my version from scratch with a little help from yours :stuck_out_tongue: 
I just finished my own recognition algorithm and now working on squashing bugs and thinking about the UI.

**Comment by flide** [comment]
Just to let everyone know, I have hired a programmer to work on this project. It would be helpful if you guys can keep pinging bugs to the repository so that we don't loose the quality.

**Comment by flide** [comment]
Hey guys iteration-3 has been released for 8vim, please checkout the new set of features now available on play store.

**Comment by reacherjack422** [comment]
Please add a donate option. Especially now that you hired a programmer.

**Comment by flide** [comment]
@reacherjack422 Thanks for your support. I am really overwhelmed to hear that you want to support this project financially. 

I am currently torn between donations and subscription based app monetization , and don't really know what is the best way to continue working on this project and other related projects like a learning game and supporting emoji keyboard. There is a lots of stuff to do, multitude of features to develop while also finding a way to support me and my family.

Any suggestion from you people would be great. I am looking for someone to guide me in this.

**Comment by reacherjack422** [comment]
@flide to be very honest, I have never paid for any subscription app, and most likely never will.

I bought the "Pro Features" for FairEmail before and I think that is definitely an acceptable model for FOSS apps. I also think a simple donate option is really nice, and my personal preference for paying for FOSS. 

In my humble opinion, anywhere less than or equal to $5 (US dollars) is probably a good price point for pro features if you go that route. Obviously, donations can be unlimited, and I think could end up being more lucrative in the long run from perhaps more generous donors.

Whatever your decide, best wishes for your success.

**Comment by giluriel** [comment]
Hi, 
These are great news!! 
Already downloded and installed. 
How can I add hebrew to the keyboard? I couldn't find settings for it
Thanks, 
Gil
--
Sent from myMail for Android יום חמישי, 01 אוקטובר 2020, 05:43אחה״צ +03:00 from Ravi Agarwal  notifications@github.com :

**Comment by flide** [comment]
@giluriel , No Hebrew translations yet, please open a new issue for the same.

**Comment by flide** [comment]
Iteration 5 out now!!! 
Grab the latest release from either play store or f-droid.

**Comment by Aq32** [comment]
I'm not sure if I should create a separate issue or use the https://github.com/flide/8VIM/issues/1 "Support Dictionary" issue but there's a feature that I would like to suggest which comes from my current main Android keyboard: [Multiling O Keyboard](https://play.google.com/store/apps/details?id=kl.ime.oh&hl=en_US&gl=US) It's not a very well documented/promoted feature but it's part of "Word prediction" and shows word suggestions in the keyboard's "Suggestions Bar" at the top. 

There's an option in the settings for "Word prediction" called: "Vowel optional". The description for it is: "Predict words without typing some of it's vowels. eg. fbls->fabulous, intrntnlatn->internationalization..."
The way this works is that you swipe a normal (qwerty) keyboard except you can skip out the vowels and in combination with "Guess missing letters" and "Next Word" options you can speed up text input significantly.

To give an example, to type: "Good morning" I would swipe the letters: g->d->m->r->n>g on the qwerty keyboard. This is significantly faster than typing each letter by push and swiping every single one of the 12 characters (including space). You can even add new words to the keyboard's dictionary so you have your own word or words with very minimal swiping movements.

Something like this is quite innovative and fast for a qwerty keyboard if you're precise in your swipes. Maybe this idea cannot be carried out conceptually as easily in the user's head for 8vim though. I think the idea of trying to make 8vim even faster is very much worth talking about even if this isn't the right way to do it. At minimum, a "Suggestions Bar" for different words from a dictionary would definitely speed up text input.

**Comment by flide** [comment]
New ideas are pouring in too fast... and quality ideas at that ... We need a better issue tracking system then the current one. Any suggestion?

**Comment by flide** [comment]
> I'm not sure if I should create a separate issue or use the https://github.com/flide/8VIM/issues/1 "Support Dictionary" issue but there's a feature that I would like to suggest which comes from my current main Android keyboard: [Multiling O Keyboard](https://play.google.com/store/apps/details?id=kl.ime.oh&hl=en_US&gl=US) It's not a very well documented/promoted feature but it's part of "Word prediction" and shows word suggestions in the keyboard's "Suggestions Bar" at the top. 
> 
> There's an option in the settings for "Word prediction" called: "Vowel optional". The description for it is: "Predict words without typing some of it's vowels. eg. fbls->fabulous, intrntnlatn->internationalization..."
> The way this works is that you swipe a normal (qwerty) keyboard except you can skip out the vowels and in combination with "Guess missing letters" and "Next Word" options you can speed up text input significantly.
> 
> To give an example, to type: "Good morning" I would swipe the letters: g->d->m->r->n>g on the qwerty keyboard. This is significantly faster than typing each letter by push and swiping every single one of the 12 characters (including space). You can even add new words to the keyboard's dictionary so you have your own word or words with very minimal swiping movements.
> 
> Something like this is quite innovative and fast for a qwerty keyboard if you're precise in your swipes. Maybe this idea cannot be carried out conceptually as easily in the user's head for 8vim though. I think the idea of trying to make 8vim even faster is very much worth talking about even if this isn't the right way to do it. At minimum, a "Suggestions Bar" for different words from a dictionary would definitely speed up text input.

This definately belongs to predictive text support, but not the current one. we will keep it here till #1 is not fixed.

**Comment by doak** [comment]
> We need a better issue tracking system then the current one. Any suggestion?

@flide, you can setup "projects" to keep track of basic development and e.g. multiple advanced topics. The you are able to use labels like "To Do", "In Progress", and "Done" to manage the related issues on a Kanban-like board.
That way you can cluster bug reports, feature requests, pull requests and even discussions and nevertheless keep track of probably most important current development.
Just my 2 cents ...

**Comment by flide** [comment]
@doak Thanks, I will look into the 'projects' thing.

Also, your 2 cents are better then my no cent.

**Comment by foolsbazar** [comment]
Hi @flide! I'm so glad you raised the 8pen concept again from the ashes. When I entered F-Droid couple of days ago I swear I lost one heartbeat so much I was happy. 
I used to develop a similar project for Sailfish OS some years ago ([sailfish-loopen](https://openrepos.net/content/foolsbazar/loopen-keyboard-fork), source code [here](https://github.com/foolsbazar/sailfish-loopen-vkb)). Actually I didn't create the original code, but then I modified and maintained it for some years myself.
I'm not a dev actually, so I'm afraid I will not be able to help you on this project (I know nothing about java), but I guess If I'll find the time I'll give it a try.
I just wanted to let you know about some features present in my app that might be of inspiration for you:
 - Possibility to insert space and backspace characters without raising the finger from the keyboard (by entering inside a sector and going back to the centre).
 - Inserting letters with accent by first typing the origin letter and then the accent (e.g. a+´ = á). The accents layout would be triggered by doing again back and forth into another sector.
 - Possibility of rewinding swipes.
 - Printing letters in advance while swiping slowly (extremely helpful for beginners).
 - Capitalize the letters if the swipe starts from a sector rather then the centre.

Some features that I wanted to insert but never did because I gave up on Sailfish OS:
 - Possibility to insert a double (or many) letter by drawing a (or many) circle in the last sector of the swipe.
 - Possibility of personalizing the letters layout.
 - Possibility of personalizing the sectors functions.

Anyway thank you for making this project available open source! Please do not move to a subscription model for this app. The learning curve for this keyboard is already too steep to attract new people, if a subscription will be required you will have to stick only with those users who already love it (of which, by the way, I'll be a part).

P.S. I didn't have time to go through the issues, but I will propose the above mentioned features as feature requests in the following days ;-)

**Comment by flide** [comment]
A release **was** planned for 13th Nov, but this is going to be delayed as it's a festive season in India right now... Too much to do, too little time... sorry folks. 

Should have forseen it...

**Comment by m-rey** [comment]
> I am currently torn between donations and subscription based app monetization

I'm also strongly for donations.

Subscriptions are nice, as they give you a steady income stream. However, I have to agree with @reacherjack422:  
Hardly anyone pays for a subscription.

Many want to support FOSS, but are not willing (or simply can't afford) to pay regularly to use an application. However, adding donations would allow for more people to give money to the project, as they can choose when and how much to donate.

**Comment by f1reflyyyylmao** [comment]
I think subscriptions could work out if this project goes the krita route: Be available like it is right now from f-droid, archive.org, own builds etc, but charge money in some form on the playstore. When krita entered the windows store it got a huge money boost, even when it remained gratis in the gnu/linux distribution repositories. The playstore users are somewhat more used to pay money for stuff and don't have any expectations regarding software freedom, which one can easily abuse as developer.

**Comment by flide** [comment]
I have a feeling that chatting would be more pleasant then posting in a forum.
I am going to try Matrix to develop a better channel of communication for the community.
Everybody is invited to join .
[8vim matrix community](https://matrix.to/#/+8vim:matrix.org)

**Comment by flide** [comment]
I just completed Iteration-6 release. Both google play and fdroid should show up the updates soon.
If anyone is impatient, they can download from the github release - https://github.com/flide/8VIM/releases/tag/Iteration-6

**Comment by Glitchy-Tozier** [comment]
@**Everyone who is interested in a new/better letter-layout for 8VIM** (English as well as new languages):

Please share what movements feel good to you HERE: https://github.com/flide/8VIM/discussions/140

**Comment by b9Joker108** [comment]
I just have a general question: how long does it generally take to become proficient and fast in typing with this app?

**Reply by geofrey** [reply]
I guess the timing would vary over individuals and circumstances but IIRC I took about a week to learn most of the letters. For me, memorization went faster when I turned off 'display letters on X-pad'. (Maybe draw out a paper copy of the layout for reference?)

**Reply by b9Joker108** [reply]
That is interesting that you recommend having an external reference and making the letters invisible. I didn't know that could be done. I will keep that in mind. I had already noticed that my fingers get in the way of viewing the key legend when I am actively typing with it. Which is frustrating and annoying. Others have noted similarly, as I have read. I think the app is novel and has potential and I can perceive its potential power in mastery. I will persevere. Does anyone actually use this app in the commandline and with Vim/Neovim? Also, I often use many different diacritics. I have heard the app supports them, is that difficult to enable?

**Reply by MaethorNaur** [reply]
There's a wiki page explaining how the keyboard works (https://github.com/8VIM/8VIM/wiki/Usability-Guide) with common movement/characters.
However, there isn't a way to know all the characters' mapping at this point (especially if you use another layout than English).
We are thinking of a way to make it easier like a tutorial/beginner mode/ map screen within the app

**Comment by b9Joker108** [comment]
To whomever it may concern

I consider that a functional and well-designed interactive and pragmatic
tutorial  learning mode, grounded in the academic discipline of Mnemonics,
with the aim to quickly upskill the new user by rote and muscle-memory,
would be an excellent addition and enhancement to the application. The
short learning session prompt of the tutor, should pop-up in the Android
notifications according to the scientifically optimised brief but
regularly re-occurring, learning session.

The interactive tutor would efficiently and effectively facilitate the
uptake, implementation and stick-ability of a new user in perservering with
the application through the not-inconsiderable period of the steep learning
curve in becoming habituated to its practical usage.

Case in point, due to the absence of this functionality, I have strayed
from the regular discipline of habituation to the gestures of the
application by rote and have reverted to generally using Typewise. *Looks
sheepish*. I do though, have all possible keyboards installed on my Android
devices and generally cycle between them for my own purposes, as none of
them fulfills all my requirements in my diverse range of digital activities.

This application is a considerable time and resource investment for the new
user, as this is functionally a non-productive period and the application
is profoundly counter-intuitive and non-standard. Having clear visual,
kinaesthetic and auditory cues to learn and efficiently and effectively
mnemonically reinforce the touchscreen gestures via a GUI tutor,
reinforcing the motion of the gestures and speed their acquisition by
memorisation, will fundamentally and profoundly improve the project and go
a considerable way, to ensuring its longevity and uptake by more mature
users in particular, as well as the young. It just makes for a better
all-round application.

Moreover, with imperative, I also recommend and foreground the iteration of
the keyboard functionality with the ability to toggle between a standard
English textual input mode and a scientifically designed input mode of an English
keyboard mapping and binding, particularly geared towards efficiently
inputting commands and text in the shell commandline and for usage with vi
and Vim, in particular.

Respectfully and with all sincerity and thanks
Beauford

PS:  I use many different Latin/Roman diacritics. I feel that these would be best implemented in this keyboard by a popup window from the root-letter keybinding, then in the popup window, the user would gesture to the particular user-specified diacritic bound to the occurrence of the root-letter or base letter without the diacritic, as is standard in a number of virtual keyboards. I hope that is clear. Even upon re-reading, it does sound rather cryptic.
